src/Microsoft.AspNetCore.Routing/RouteValuesBasedEndpointFinder.cs

```diff
@@ -45,11 +45,9 @@ public IEnumerable<Endpoint> FindEndpoints(RouteValuesBasedEndpointFinderContext
                     context.ExplicitValues,
                     context.AmbientValues);
             }
-            else if (_namedMatches.TryGetValue(context.RouteName, out var linkGenerationTree))
+            else if (_namedMatchResults.TryGetValue(context.RouteName, out var namedMatchResults))
             {
-                matchResults = linkGenerationTree.GetMatches(
-                    context.ExplicitValues,
-                    context.AmbientValues);
+                matchResults = namedMatchResults;
```

Comment:
I wanted to write a useful comment here as to why it is this way but felt it's going to need a good big explanation (which probably we want to avoid in a comment?)

Reply:
Explanation for the change:
Decision tree is built using the 'required values' of actions.
- When generating a url using route values, we check the explicitly supplied route values + ambient values to see if they match the required values based tree.
- When generating a url using route name, route values for controller, action etc. might not be provided (this is expected because as a user i want to avoid writing all those and instead chose to use a route name which is quick). So since these values are not provided and might not be even in ambient values, we decision tree would fail to find a match.

Reply:
I think this is a good explanation, and importantly it's not too big to leave in a comment. I'd suggest a big comment at the beginning of the method body that explains what cases matter and why they are different. examples go a looooong way.